src/server.rs

```diff
@@ -120,11 +124,20 @@ impl Service for ServerHandler {
     type Error = HyperError;
     type Future = ServerHandlerFuture;
 
+    // TODO make the parameter name configurable so there are no collisions with the actual server to be stubbed.
     fn call(&mut self, req: HyperRequest<Body>) -> <Self as Service>::Future {
         let auto_cors = self.auto_cors;
         let sources = self.sources.clone();
-        let provider_state = self.provider_state.clone();
+        let mut provider_state = self.provider_state.clone();
         let (parts, body) = req.into_parts();
+        {
+            let yeah = &parts;
+            let provider_state_header = yeah.headers.get(self.provider_state_header_name.clone());
+            if provider_state_header.is_some() {
+                provider_state = Some(Regex::new(provider_state_header.unwrap().to_str().unwrap()).unwrap());
```

Comment:
Lots of unwrapping here. Better to use pattern matching so we know there won't be any panics. 

You can also use an `if let Some(header) = provider_state_header` instead of the `if provider_state_header.is_some()` followed by the later unwrap.